Segmented scan agent with multiple segments per block

oleksandr-pavlyk opened the pull request:
## Description

<!-- Every PR should have a corresponding issue that describes and motivates the work done in the PR -->
Closes #6783 
Closes #8424

  - extend `agent_segmented_scan` with ability to to consume multiple segments per block
     - Introduce new agent template parameter `MaxSegmentsPerBlock` which controls the size of `OffsetT` array in shared memory dedicated to storing inclusive-scan of segment sizes. The actual number of segments to process is passed dynamically to `consume_ranges` method. It can not exceed `MaxSegmentsPerBlock`, however.
     - Multiple-segments are processed by reconstructing head-flags while loading values using `multi_segment_iterator` class, which is read- and write- transform iterator.  
        - Read transformation reconstructs head-flag and forms value-flag pair, used as `AccumT` by `BlockScan`, and  
        - write-transform unpacks value-flag pair and only writes the value to the output iterator.
     - Mapping from `element_id` to `segment_id` and `relative_offset` (or rank) is encapsulated into searcher types.
     - Three searcher types are implemented:
        - When segment sizes are all the same, `segment_id = element_id / segment_size`.
        - Generically, binary search in the array of inclusive-scan of segment sizes is performed. Two implementations (types) are provided:
           - When static bound on array size of known (as it is in our case), branch-less binary search provides best performance
           - When this bound is unknown, `cuda::std::upper_bound` can be used for larger number of sizes, and linear search using dynamic for-loop for  smaller sizes
     - While computing inclusive-scan of segment sizes, we check if all sizes to be processed by block are the same, and use fixed-size searcher, otherwise use branch-less generic searcher.
     - When dynamic number of segments equals 1, simpler `consume_range` call is made, which does not reconstruct head-flags 

  - introduce `agent_warp_segmented_scan` which closely follows the logic of `agent_segmented_scan` that computes are block-granularity, but using warps to process given number of segments
    - This agent is advantageous when segment sizes are medium-sized, but not too small.
  - introduce `agent_thread_segmented_scan` which implement thread-granularity work. Each thread processes requested number of segments.
     - multi-segment iterator here is implemented sequentially, iterating within segment, then over segments.
     - Thread-granularity agent is ideal when segments have only a few elements each.
   - Dispatch layer is modified to take 2 additional parameters (with default values), the number of segments to be processed by a worker (defaults to 1), and the type of worker requested (defaults to block).

<hr>

<strike>
This PR implements `cub::detail::segmented_scan::agent_segmented_scan_multiple_segments_per_block`.
The agent processes statically known `int SegmentsPerBlock` segments per block in `consume_ranges` method, and implements a tail version as well. 
</strike>

<!-- Provide a standalone description of changes in this PR. -->

<!-- Note: The pull request title will be included in the CHANGELOG. -->

## Checklist
<!-- TODO: - [ ] I am familiar with the [Contributing Guidelines](). -->
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

oleksandr-pavlyk commented:
/ok to test

oleksandr-pavlyk commented:
/ok to test

oleksandr-pavlyk commented:
/ok to test

oleksandr-pavlyk commented:
/ok to test

bernhardmgruber commented on cub/benchmarks/bench/segmented_scan/base.cuh:
```diff
@@ -0,0 +1,300 @@
+  return bench_impl<1, T, OffsetT>(state, tl);
+}
+
+#if !(_CCCL_CUDA_COMPILER(NVCC, ==, 12, 0))
```
```suggestion
#if _CCCL_CUDA_COMPILER(NVCC, >=, 12, 1)
```

bernhardmgruber commented on cub/benchmarks/bench/segmented_scan/base.cuh:
```diff
@@ -0,0 +1,300 @@
+using benched_value_types = all_types;
+#else
+// WAR for excessive time CTK 12.0 CICC takes to compile these benchmarks for int128_t
+using benched_value_types = nvbench::type_list<int8_t, int16_t, int32_t, int64_t, float, double, complex>;
```
Please apply the optimization described here: https://nvidia.github.io/cccl/unstable/cub/tuning.html#nvbench-attributes, even if we probably never run the benchmarks with nvcc 12.0.

Alternatively, we can just disable the benchmark entirely for nvcc 12.0 (e.g. by wrapping a big `#if/#else` around it

bernhardmgruber commented on cub/benchmarks/bench/segmented_scan/sum.cu:
```diff
@@ -0,0 +1,15 @@
+// SPDX-FileCopyrightText: Copyright (c) 2011-2023, NVIDIA CORPORATION. All rights reserved.
+// SPDX-License-Identifier: BSD-3
```
Important: Any new file added to CCCL should have the following license:

```suggestion
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
Comment applies to all new files.

bernhardmgruber commented on cub/benchmarks/bench/segmented_scan/base.cuh:
```diff
@@ -0,0 +1,300 @@
+  };
+
+  using MaxPolicy = policy_t;
+};
```
Important: Policy hubs are phased out and should replaced by policy selectors. We translated most of CUB already, so you should fined ample examples. AI is also good at making this translation.

If you need more background, here is the design document: https://docs.google.com/document/d/1KtrSsMk2LAXSB5cdAUnmn3BMow5y6B8BhC5-XVA434g

bernhardmgruber commented on cub/benchmarks/bench/segmented_scan/base.cuh:
```diff
@@ -0,0 +1,300 @@
+  summary.set_int64("value", num_segments);
+
+  thrust::device_vector<T> input = generate(elements);
+  thrust::device_vector<T> output(elements);
```
```suggestion
  thrust::device_vector<T> output(elements, thrust::no_init);
```

oleksandr-pavlyk replied:
Using `no_init` tag requires `T` to be trivially constructible, and `T=complex` is not.  So my choices are to either be paying the price of performing unnecessary initialization for all types, or create utility and place it into `nvbench_helper`, or make it available even broader as this issue often arises in tests as well

```cpp
template <typename T, typename = void>
auto make_output_device_vector(cuda::std::size_t elements) {
   return thrust::device_vector<T>(elements);
}

template <typename T, std::enable_if_t<cuda::std::is_trivially_constructible_v<T>, void>>
auto make_output_device_vector(cuda::std::size_t elements) {
   return thrust::device_vector<T>(elements, thrust::no_init);
}
```

What do you think?

bernhardmgruber replied:
Ah, I see. That's what I built `thrust::default_init` for.

bernhardmgruber commented on cub/benchmarks/bench/segmented_scan/base.cuh:
```diff
@@ -0,0 +1,300 @@
+  thrust::device_vector<offset_t> offsets(num_segments + 1, thrust::no_init);
+  thrust::tabulate(offsets.begin(), offsets.end(), to_offsets_functor<offset_t>{elements, segment_size, Wobble});
+
+  T* d_input          = thrust::raw_pointer_cast(input.data());
```
```suggestion
  const T* d_input          = thrust::raw_pointer_cast(input.data());
```

bernhardmgruber commented on cub/cub/agent/agent_segmented_scan.cuh:
```diff
@@ -17,13 +17,21 @@
 #include <cub/iterator/cache_modified_input_iterator.cuh>
+#include <cub/util_arch.cuh> // MemBoundScaling
 
+#include <cuda/__cmath/ceil_div.h>
```
I don't have strong opinions on this. I think the split of dispatch + kernel makes somewhat sense, but I don't see a good reason to split off agents from kernels.

oleksandr-pavlyk replied:
Right now, there are 3 agent implementation header files, and one kernel header file. I think the right split is to create 3 kernel header file, and move agent definition into respective kernel header files. 

If you agree to the split, I will reorganize sources

oleksandr-pavlyk replied:
@bernhardmgruber @miscco `"cub/agent/agent_*segmented_scan.cuh"` files each define agent and agent policy. The tuning header includes agent header for the policy definition, and kernel header includes it for the agent definition. 

By moving agent definition into kernel file we reduce compilation time, but avoiding compiling agent definition twice (once compiling tuning, and once compiling kernel).

We would still need to keep dedicated files for agent policy structs though.

oleksandr-pavlyk replied:
I moved definition of agents in the same translation units that defined the kernel that uses the agent.

bernhardmgruber commented on cub/cub/agent/agent_segmented_scan.cuh:
```diff
@@ -129,25 +150,69 @@ struct agent_segmented_scan
+  using block_reduce_t      = BlockReduce<unsigned int, block_threads>;
 
-  union _TempStorage
+  using _multiple_segment_algorithms_storage_t = union
```
Q: What's the advantage of an alias to an unnamed union vs. naming the union directly?
```suggestion
  union _multiple_segment_algorithms_storage_t
```

bernhardmgruber commented on cub/cub/agent/agent_thread_segmented_scan.cuh:
```diff
@@ -0,0 +1,488 @@
+struct agent_thread_segmented_scan_policy_t : ScalingType
+{
+  static constexpr CacheLoadModifier load_modifier = LoadModifier;
+};
```
Important: Please just define a simple struct as tuning policy, as most other CUB algorithms do by now. The struct should already contain mem-bound scaled values. The scaling should be done in the default policy selector.

bernhardmgruber commented on cub/cub/device/dispatch/tuning/tuning_segmented_scan.cuh:
```diff
@@ -30,6 +32,142 @@ CUB_NAMESPACE_BEGIN
+  {
+    return 1;
+  }
+};
```
Important: Since there is no CCCL.C exposure for the new segmented scan, all of those policy wrappers should not be needed. Furthermore, with the new tuning API design (policy selector-based), they should also not be needed anymore. Please refactor them away when migrating to policy selectors.

bernhardmgruber reviewed (commented):
Here are some comments from a quick look. I think the first thing to attach is to migrate to the new tuning API design, which most of CUB has by now. Since we already have a segmented scan implementation, that is not ported yet, you could port that yourself or let me port it. You can also try to integrate this PR using the old policy hub style and then migrate in a follow-up PR. What's important to me is that we finish all this work in time before CCCL 3.4 branches, which is in 3-4 weeks. Let me know what you think!

bernhardmgruber commented on cub/test/catch2_test_device_segmented_scan_multi_segment.cu:
```diff
@@ -648,7 +648,8 @@ std::tuple<std::vector<OffsetT>, std::vector<OffsetT>> make_in_out_offsets(const
   std::vector<OffsetT> offsets_with_gaps;
   offsets_with_gaps.resize(segment_count + 1);
   offsets_with_gaps[0] = OffsetT{0};
-  std::inclusive_scan(sizes_with_gaps.begin(), sizes_with_gaps.end(), offsets_with_gaps.begin() + 1);
```
@miscco I think we need to disable the exec check for all our STL algorithms

oleksandr-pavlyk replied:
Opened https://github.com/NVIDIA/cccl/issues/8469 to track the issue

davebayer replied:
@bernhardmgruber it's even worse, I think even incrementing a host-only iterator from `__host__ __device__` function is problematic

oleksandr-pavlyk commented:
/ok to test 184201944a02eb7b7affde72c8d936776376ddb9

oleksandr-pavlyk commented:
/ok to test c2e6729c8303a429f87022a76f5de3c571f869c5

oleksandr-pavlyk commented:
/ok to test  6087434

oleksandr-pavlyk commented:
/ok to test 48cf816

oleksandr-pavlyk commented on cub/cub/agent/agent_segmented_scan.cuh:
```diff
@@ -17,13 +17,21 @@
 #include <cub/block/block_store.cuh>
+#include <cub/detail/segmented_scan_helpers.cuh>
 #include <cub/iterator/cache_modified_input_iterator.cuh>
+#include <cub/util_arch.cuh> // MemBoundScaling
```
MemBoundScaling is no longer used in this file. Remove the include

oleksandr-pavlyk commented:
/ok to test 79240e7

oleksandr-pavlyk commented:
/ok to test df97680e23d8b712691d662c3cbf8ce49a810e64

oleksandr-pavlyk commented:
This PR now uses new tuning API. 

Thank you @bernhardmgruber 🙏 for guidance with this transition.

bernhardmgruber commented on cub/cub/device/dispatch/dispatch_segmented_scan.cuh:
```diff
@@ -66,213 +76,253 @@ struct device_segmented_scan_kernel_source
-  /// Iterator to the input sequence of data items
-  InputIteratorT d_in;
+  return dispatch_arch(policy_selector, arch_id, [&](auto policy_getter) {
+    const segmented_scan_policy active_policy = policy_getter();
```
Important: If you never need the tuning policy at compile-time in host code, you should just query it directly like:
```suggestion
    const segmented_scan_policy active_policy = policy_selector(arch_id);
```
Then you don't need to call `dispatch_arch` and save some compile-time.

oleksandr-pavlyk commented on cub/cub/device/dispatch/dispatch_segmented_scan.cuh:
```diff
@@ -15,16 +15,19 @@
 #include <cub/util_debug.cuh>
 #include <cub/util_device.cuh>
 #include <cub/util_math.cuh>
+#include <cub/util_temporary_storage.cuh>
```
Neither `"cub/util_temporary_storage.cuh"` nor `"cub/util_math.cuh"` are necessary in this translation unit.

bernhardmgruber commented on cub/cub/device/dispatch/tuning/tuning_segmented_scan.cuh:
```diff
@@ -13,95 +13,298 @@
-  {}
+  int block_threads;
+  int items_per_thread;
+  CacheLoadModifier load_modifier;
+  BlockLoadAlgorithm load_algorithm;
```
Important: Please swap those two members for consistency with other tuning policies.

bernhardmgruber commented on cub/cub/device/dispatch/tuning/tuning_segmented_scan.cuh:
```diff
@@ -13,95 +13,298 @@
+  CacheLoadModifier load_modifier;
+  WarpLoadAlgorithm load_algorithm;
+  WarpStoreAlgorithm store_algorithm;
+  int max_segments_per_warp;
```
Q: Should we rename this to just `max_segments` here and in `block_segmented_scan_policy`?

bernhardmgruber commented on cub/cub/device/dispatch/tuning/tuning_segmented_scan.cuh:
```diff
@@ -13,95 +13,298 @@
+    constexpr auto scan_transposed_blockload =
+      large_values ? BLOCK_LOAD_WARP_TRANSPOSE_TIMESLICED : BLOCK_LOAD_WARP_TRANSPOSE;
+    constexpr auto scan_transposed_blockstore =
+      large_values ? BLOCK_STORE_WARP_TRANSPOSE_TIMESLICED : BLOCK_STORE_WARP_TRANSPOSE;
+
+    constexpr int nominal_block_threads    = 128;
+    constexpr int nominal_items_per_thread = 9;
+    constexpr int max_segments_per_block   = 512;
+    constexpr int max_segments_per_warp    = 128;
+
+    // Actual ComputeT is AccumT for single segment per worker,
+    // but becomes tuple<AccumT, bool> for multi-segment per worker
+    using block_compute_t = agent_block_segmented_scan_compute_t<AccumT, max_segments_per_block>;
+    using warp_compute_t  = agent_warp_segmented_scan_compute_t<AccumT, max_segments_per_warp>;
+
+    constexpr auto block_scaled =
+      scale_mem_bound(nominal_block_threads, nominal_items_per_thread, size_as_int_v<block_compute_t>);
+    constexpr auto warp_scaled =
+      scale_mem_bound(nominal_block_threads, nominal_items_per_thread, size_as_int_v<warp_compute_t>);
+    constexpr auto thread_scaled =
+      scale_mem_bound(nominal_block_threads, nominal_items_per_thread, size_as_int_v<AccumT>);
+
+    return segmented_scan_policy{
+      block_segmented_scan_policy{
+        block_scaled.block_threads,
+        block_scaled.items_per_thread,
+        LOAD_DEFAULT,
+        scan_transposed_blockload,
+        scan_transposed_blockstore,
+        BLOCK_SCAN_WARP_SCANS,
+        max_segments_per_block},
+      warp_segmented_scan_policy{
+        warp_scaled.block_threads,
+        warp_scaled.items_per_thread,
+        LOAD_DEFAULT,
+        WARP_LOAD_TRANSPOSE,
+        WARP_STORE_TRANSPOSE,
+        max_segments_per_warp},
+      thread_segmented_scan_policy{thread_scaled.block_threads, thread_scaled.items_per_thread, LOAD_DEFAULT}};
+  }
```
Important: let's avoid repetition:
```suggestion
  [[nodiscard]] _CCCL_API constexpr auto operator()(::cuda::arch_id arch) const -> segmented_scan_policy
  {
    return policy_selector{int{sizeof(AccumT)}}(arch);
  }
```

oleksandr-pavlyk commented:
I merged main to pick up env overloads added by @gonidelis in #8610 . I resolved conflicts and updated calls to `dispatch` since they changed in this PR.

oleksandr-pavlyk committed:
Add thread worker

NaderAlAwar commented on cub/cub/device/dispatch/kernels/kernel_warp_segmented_scan.cuh:
```diff
@@ -0,0 +1,664 @@
+    _CCCL_ASSERT(n_segments <= max_segments, "Number of segments should not exceed statically provisioned storage");
+
+    // cooperatively compute inclusive scan of sizes of segments to be processed by this block
+    {
```
Suggestion: if possible, move this block of code to a separate device function, maybe named something like `compute_segment_offsets_and_fixed_size_mask`

oleksandr-pavlyk committed:
Add missing include

Include is needed to define c2h::type_list

NaderAlAwar commented on cub/benchmarks/bench/segmented_scan/base.cuh:
Suggestion: you probably want to add `"warp"` and `"thread"`

oleksandr-pavlyk replied:
Using worker "thread" for larger segment size 4417 would be quite slow, so I did not work to exercise it by default.

oleksandr-pavlyk committed:
Use different workers for different segment sizes using 3-way partition

Add a test that partitions input segments into small/medium/large and
invokes segmented scan with different policies, different workers and
different number of segments per worker according to the category.

NaderAlAwar commented on cub/cub/device/dispatch/kernels/kernel_thread_segmented_scan.cuh:
```diff
@@ -0,0 +1,542 @@
+  }
+
+public:
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE multi_segmented_seq_iterator(
```
Important: this should likely be `_CCCL_DEVICE` only, same for the other member functions. It also calls `set_first_nonempty`, which is device only

oleksandr-pavlyk committed:
Remove unneccessary include

oleksandr-pavlyk committed:
Factored out scan_segments_chunked, and process_segment_sizes

oleksandr-pavlyk committed:
Refactored scan_single_segment_chunked to helper header

Implemented scope structs that plug into that helper and used
it in agent_segmented_scan and in agent_warp_segmented_scan

oleksandr-pavlyk committed:
Refectored selection of segmented scan searcher

oleksandr-pavlyk committed:
Scope adapter cleanup

Consistent naming of helper scope classes.
Cleanup.

oleksandr-pavlyk committed:
Use single_segment_scan_chunked in thread-level agent

oleksandr-pavlyk committed:
Consolidation of scope structs into single scope

oleksandr-pavlyk committed:
Removed includes rendered unnecessary by refactoring

oleksandr-pavlyk committed:
Properly implement testing of multi-segment dispatch for host-, device- and graph- launch

bernhardmgruber commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Broke up _CCCL_ASSERT on load_modifier to switch on worker enumeration value

oleksandr-pavlyk committed:
Add comments making some assumptions explicit

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Remove explicit constructor for worker_prefix_callback_t struct, use aggregate initialization instead

oleksandr-pavlyk committed:
Add a comment that for thread worker num_segments_per_worker does not have upper bound to enforce

oleksandr-pavlyk committed:
Clarified assertion message in select_segment_scan_searcher

oleksandr-pavlyk committed:
Refined comment in the body of device_warp_segmented_scan_kernel to make correctness assumption clear

oleksandr-pavlyk committed:
Correct static assertion message text to use current method name

oleksandr-pavlyk committed:
Added comment that thread-level agent does not temp-storage per PR feedback

oleksandr-pavlyk committed:
Explicitly set m_segment_id to m_num_segments if set_first_nonempty does not find a non-empty segment

oleksandr-pavlyk committed:
Replaced all _CCCL_HOST_DEVICE in helper to _CCCL_DEVICE

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Add comment for clarity

oleksandr-pavlyk committed:
Clamp num_segments_per_worker for thread-agent to ensure that segmented_id * block_size is representable by int

oleksandr-pavlyk committed:
Avoid using device code lambdas per coding guideline

oleksandr-pavlyk committed:
Address coding guideline issues

oleksandr-pavlyk committed:
Fix grammer in comments

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Add missing includes for IWYU

oleksandr-pavlyk committed:
Thread agent optimization

Exit early if remaining work is out of bound. Clamp amount
of the segments for this thread to process to avoid probing
out of bound.

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk commented on cub/cub/device/dispatch/tuning/tuning_segmented_scan.cuh:
```diff
@@ -105,33 +174,46 @@ struct policy_selector
+        LOAD_DEFAULT,
+        WARP_STORE_TRANSPOSE,
+        max_segments_per_warp},
+      thread_segmented_scan_policy{thread_scaled.threads_per_block, thread_scaled.items_per_thread, LOAD_DEFAULT}};
```
🤷🏻. I do not see a problem here. Algorithm tuning will be performed before the algorithms moves out of `detail` namespace. But this is for a follow up future PR

oleksandr-pavlyk commented on cub/cub/device/dispatch/tuning/tuning_segmented_scan.cuh:
```diff
@@ -105,33 +174,46 @@ struct policy_selector
+        LOAD_DEFAULT,
+        WARP_STORE_TRANSPOSE,
+        max_segments_per_warp},
+      thread_segmented_scan_policy{thread_scaled.threads_per_block, thread_scaled.items_per_thread, LOAD_DEFAULT}};
```
I do not see any issue with this tuning. I am going to add a comment to the code to make it clear.

oleksandr-pavlyk committed:
Add comment for max_segments_per_warp in tuning_segmented_scan

This is to make it clear that max_segments_per_warp is the natural
kernel tuning parameter. It is independent from max_segments_per_block.

oleksandr-pavlyk commented:
@coderabbitai review

oleksandr-pavlyk committed:
Address code-rabbit's nitpick

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Fix comment typo

oleksandr-pavlyk committed:
Addressing review feedback

1. Use _CCCL_DEDUCTION_GUIDE_ATTRIBUTES instead of _CCCL_DEVICE
2. Avoid allocating __shared__ temp-storage of empty type.
3. Add assertion that access to input/output offsets in multi-segment
   sequential iterator is done on a non-exhausted iterator.

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Add missing cuda/sts/__algorithm/max.h include

oleksandr-pavlyk commented:
@coderabbitai full review

oleksandr-pavlyk committed:
Add warp worker

oleksandr-pavlyk committed:
Fix pre-commit
